File: ButtonCore/src/main/java/buttondevteam/lib/chat/ChatMessage.java
package buttondevteam.lib.chat;

import buttondevteam.lib.player.ChromaGamerBase;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import org.bukkit.command.CommandSender;

@Builder
@Getter
public class ChatMessage {
	/**
	 * The sender which sends the message.
	 */
	private final CommandSender sender;
	/**
	 * The Chroma user which sends the message.
	 */
	private final ChromaGamerBase user;
	/**
	 * The message to send as the user.
	 */
	@Setter
	private String message;
	/**
	 * Indicates whether the message comes from running a command (like /tableflip). Implemented to be used from Discord.
	 */
	private boolean fromCommand;
	/**
	 * The sender which we should check for permissions. Same as {@link #sender} by default.
	 */
	private CommandSender permCheck;
	/**
	 * The origin of the message, "Minecraft" or "Discord" for example. May be displayed to the user.<br>
	 *     <b>This is the user class capitalized folder name.</b>
	 */
	private final String origin;

	/**
	 * The sender which we should check for permissions. Same as {@link #sender} by default.
	 *
	 * @return The perm check or the sender
	 */
	public CommandSender getPermCheck() {
		return permCheck == null ? sender : permCheck;
	}

	private static ChatMessageBuilder builder() {
		return new ChatMessageBuilder();
	}

	@NonNull
	public static ChatMessageBuilder builder(CommandSender sender, ChromaGamerBase user, String message) {
		return builder().sender(sender).user(user).message(message).origin(user.getFolder().substring(0, 1).toUpperCase() + user.getFolder().substring(1));
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/chat/ChatMessage.kt
package buttondevteam.lib.chat

import buttondevteam.lib.player.ChromaGamerBase
import java.util.*

class ChatMessage internal constructor(
    /**
     * The Chroma user who sent the message.
     */
    val user: ChromaGamerBase,
    /**
     * The message to send as the user.
     */
    var message: String,
    /**
     * Indicates whether the message comes from running a command (like /tableflip). Implemented to be used from Discord.
     */
    val isFromCommand: Boolean,
    /**
     * The sender which we should check for permissions. Same as [.sender] by default.
     */
    val permCheck: ChromaGamerBase,
    /**
     * The origin of the message, "Minecraft" or "Discord" for example. May be displayed to the user.
     *
     * **This is the user class capitalized folder name by default.**
     */
    val origin: String
) {

    class ChatMessageBuilder internal constructor(
        private var user: ChromaGamerBase,
        private var message: String,
        private var origin: String
    ) {
        private var fromCommand = false
        private var permCheck: ChromaGamerBase? = null

        fun fromCommand(fromCommand: Boolean): ChatMessageBuilder {
            this.fromCommand = fromCommand
            return this
        }

        fun permCheck(permCheck: ChromaGamerBase): ChatMessageBuilder {
            this.permCheck = permCheck
            return this
        }

        fun origin(origin: String): ChatMessageBuilder {
            this.origin = origin
            return this
        }

        fun build(): ChatMessage {
            return ChatMessage(user, message, fromCommand, permCheck ?: user, origin)
        }

        override fun toString(): String {
            return "ChatMessage.ChatMessageBuilder(user=$user, message=$message, fromCommand=$fromCommand, permCheck=$permCheck, origin=$origin)"
        }
    }

    companion object {
        @JvmStatic
        fun builder(user: ChromaGamerBase, message: String): ChatMessageBuilder {
            return ChatMessageBuilder(
                user, message,
                user.folder.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
            )
        }
    }
}